### PR TITLE
Object permissions mechanism + regions (for BO Asset, Licences and Supports)

### PR DESCRIPTION
**Pull request by mkurek**
connected to #1452 
- object permissions mechanism (based on currently logged-in user)
- permissions are applied in django admin list and single object edit form
- user_permission decorator for functions to check if user has permission to object
- permissions to objects usage as regions in BO Asset, Licence and Support

### CODE REVIEW

File: src/ralph/lib/permissions/models.py [hunk 1]
```diff
@@ -25,28 +27,98 @@ def get_perm_key(action, class_name, field_name):
     return '{}_{}_{}_field'.format(action, class_name, field_name)
 
 
-class PermissionByFieldBase(ModelBase):
+class user_permission(object):
```
Comment (ar4s) [inline src/ralph/lib/permissions/models.py]: add `# noqa`

File: src/ralph/accounts/models.py [hunk 2]
```diff
@@ -5,6 +5,35 @@
...
+    class Permissions:
+        has_access = has_region
+
+
+@user_permission
+def object_has_region(user):
+    return models.Q(region__in=user.regions_ids)
+
+
+class Regionalizable(PermForObjectMixin):
```
Comment (mkurek) [inline src/ralph/accounts/models.py]: add docstring

File: src/ralph/accounts/models.py [hunk 3]
```diff
@@ -5,6 +5,35 @@
...
+
+
+class Region(PermForObjectMixin, NamedMixin):
+    """Used for distinguishing the origin of the object by region"""
+    class Permissions:
+        has_access = has_region
+
+
+@user_permission
+def object_has_region(user):
```
Comment (mkurek) [inline src/ralph/accounts/models.py]: add docstring

File: src/ralph/lib/permissions/admin.py [hunk 4]
```diff
@@ -53,3 +55,61 @@ def get_list_display(self, request):
...
+    `ralph.lib.permissions.models.PermForObjectMixin`
+    """
+    def _check_obj_permission(self, request, obj=None):
+        """
+        Returns True if user has access to object.
+        """
+        obj_permission = True
+        if obj and isinstance(obj, PermForObjectMixin):
+            obj_permission = obj.has_permission_to_object(request.user)
+        # TODO? check related objects permissions?
```
Comment (mkurek) [inline src/ralph/lib/permissions/admin.py]: remove this line

File: src/ralph/lib/permissions/admin.py [hunk 5]
```diff
@@ -53,3 +55,61 @@ def get_list_display(self, request):
         ]
 
         return list_display or ['__str__']
+
+
+class PermissionPerObjectAdminMixin(object):
```
Comment (mkurek) [inline src/ralph/lib/permissions/admin.py:91]: TODO: save validation for related objects

File: src/ralph/lib/permissions/tests/models.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+    def __str__(self):
+        return self.title
+
+
+class LongArticle(Article):
+    remarks = models.CharField(max_length=100)
+
+    class Permissions:
+        has_access = has_long_title
```
Comment (mkurek) [inline src/ralph/lib/permissions/tests/models.py:43]: add comment that this permission is anded (&) with ancestor class permissions

File: src/ralph/lib/permissions/tests/test_permissions_per_object.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+            self.long_article.has_permission_to_object(self.user1)
+        )
+
+    def test_author_should_not_have_access_to_long_article_when_it_is_not_long(self):  # noqa
+        self.assertFalse(
+            self.long_article_2.has_permission_to_object(self.user1)
+        )
+
+    def test_get_objects_for_user(self):
+        # TODO: ddt
```
Comment (mkurek) [inline src/ralph/lib/permissions/tests/test_permissions_per_object.py]: apply TODO

**Comment by ar4s** [comment]
:ok:

File: src/ralph/lib/permissions/tests/test_permissions_per_object.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+            content='lorem ipsum'
+        )
+
+
+@ddt
+class TestPermissionsPerObject(PermissionsPerObjectTestMixin, TestCase):
+    def setUp(self):
+        self._init_articles()
+
+    def test_author_should_have_access_to_his_article(self):
```
Comment (vi4m) [inline src/ralph/lib/permissions/tests/test_permissions_per_object.py:118]: Lovely.

File: src/ralph/lib/permissions/tests/test_permissions_per_object.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+        self.assertEqual(
+            str(result),
+            "(OR: ('author', {u}), ('collaborators', {u}))".format(
+                u=repr(self.superuser)
+            )
+        )
+
+
+class PermissionsPerObjectTestMixin(object):
+    def _init_articles(self):
```
Comment (vi4m) [inline src/ralph/lib/permissions/tests/test_permissions_per_object.py]: this is rather some kind of factory, init doesn't create objects, factory does.

**Comment by szok** [comment]
:ok:

**Comment by ar4s** [comment]
:ok:

**Pull request merged**